File: 2024/day19/Python/day19.py
from time import time
from functools import cache

time_start = time()

#Input
INPUT_FILE = "day19_input.txt"
blocks = [block.splitlines() for block in open(INPUT_FILE, "r").read().split("\n\n")]

available_towel_patterns = list(blocks[0][0].split(", "))

# Memoized function to count the number of ways to decompose the designs
@cache
def count_ways(designs):
    if not designs:
        return 1
    return sum(count_ways(designs[len(p):]) for p in available_towel_patterns if designs.startswith(p))

#solving
solution1, solution2 = 0, 0
for designs in blocks[1]:
    solution1 += 1 if count_ways(designs) else 0
    solution2 += count_ways(designs)

#printing solutions
print(f"Part 1: {solution1}  ({time() - time_start:.3f}s)")
print(f"Part 2: {solution2}  ({time() - time_start:.3f}s)")

File: 2024/day19/R/day19.R

# File with Santa's towel rules
file_name <- 'day19_input.txt'

# Load input data
input_data <- readLines(file_name)
towel_patterns <- strsplit(input_data[1], ", ")[[1]]
mystery_patterns <- input_data[-(1:2)]

# Holiday cache for pattern matching
cache <- collections::dict()

calculate_possibilities <- function(towels, pattern) {
  if (cache$has(pattern)) {
    return(cache$get(pattern))
  }
  
  if (pattern == "") return(1)
  
  possible_count <- 0
  for (towel in towels) {
    if (startsWith(pattern, towel)) {
      possible_count <- possible_count + calculate_possibilities(towels, gsub(paste0("^", towel), "", pattern))
    }
  }
  
  cache$set(pattern, possible_count)
  return(possible_count)
}

# Calculate results for all mystery patterns
results <- sapply(mystery_patterns, function(x) calculate_possibilities(towel_patterns, x))

# Part 1: Count valid patterns
valid_patterns <- sum(results > 0)
cat("Part 1: Number of valid patterns:", valid_patterns, "\n")

# Part 2: Total possible combinations
total_combinations <- sum(results)
cat("Part 2: Total combinations:", total_combinations, "\n")